File: src/1_100/0008_String_to_Integer_(atoi)/String_to_Integer_(atoi).java
class Solution {
    public int myAtoi(String str) {
        if(str==null || str.length()==0) {
            return 0;
        }
        java.math.BigDecimal min = new java.math.BigDecimal(Integer.MIN_VALUE);
        java.math.BigDecimal max = new java.math.BigDecimal(Integer.MAX_VALUE);
        int max_len = (""+Integer.MAX_VALUE).length()+1;
        StringBuilder sb = new StringBuilder();
        str = str.trim();

        int index = 0;
        if(str.length() > 0) {
            if(str.startsWith("-")) {
                sb.append("-");
                index++;
            }
            else if(str.startsWith("+")) {
                index++;
            }
        } else {
            return 0;
        }

        for(;index<str.length();index++) {
            char c = str.charAt(index);
            if(c==' ') {
                break;
            }else if(c=='.') {
                break;
            }else if(c>='0' && c<='9') {
                sb.append(c);
            }else {
                break;
            }
        }


        long lo = 0L;
        java.math.BigDecimal b = null;
        try {
            b = new java.math.BigDecimal(sb.toString());
            if(b.compareTo(min) < 0) {
                lo = Integer.MIN_VALUE;
                return (int)lo;
            }
            if(b.compareTo(max) > 0) {
                lo = Integer.MAX_VALUE;
                return (int)lo;
            }
            lo = Long.parseLong(sb.toString());
        }catch(Exception e) {
            return 0;
        }
        return (int)lo;
    }
	
	//这题非常坑，各种边界问题需要特别注意
	//这里给出测试用例
    public void go() {
        System.out.println("-+1 ->"+myAtoi("-+1"));
        System.out.println("20000000000000000000 ->"+myAtoi("20000000000000000000") );
        System.out.println("42 ->"+myAtoi("42"));
        System.out.println("   -42 ->"+myAtoi("   -42"));
        System.out.println("4193 with words ->"+myAtoi("4193 with words"));
        System.out.println("words and 987 ->"+myAtoi("words and 987"));
        System.out.println("-91283472332 ->"+myAtoi("-91283472332"));
        System.out.println("1234567890 ->"+myAtoi("1234567890"));
        System.out.println("+-2 ->"+myAtoi("+-2"));
        System.out.println("  -0012a42 ->"+myAtoi("  -0012a42"));
        System.out.println("3.14159 ->"+myAtoi("3.14159"));
        System.out.println("  -0012a42 ->" + myAtoi("  -0012a42"));
        System.out.println("-13+8 ->" + myAtoi("-13+8"));
    }
	
	
	
}

File: src/1_100/0008_String_to_Integer_(atoi)/String_to_Integer_(atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        if(str==None or len(str)==0 or str==""):
            return 0
        d = {'0':0, "1":1, "2":2, "3":3, "4":4, "5":5, "6":6, "7":7, "8":8, "9":9}
        i = 0
        size = len(str)
        isNegative = False
        res = 0
        max_value = 2**31-1
        min_value = -(2**31)
        while(i<size and str[i]==' '):
            i += 1
        if(i<size and str[i]=='-'):
            isNegative = True
            i += 1
        elif(i<size and str[i]=='+'):
            i += 1
        elif(i<size and (str[i]<'0' or str[i]>'9')):
            return 0
        while(i<size and str[i]>='0' and str[i]<='9'):
            res = res*10 + d[str[i]]
            if(res > max_value and not isNegative):
                return max_value
            if(-res < min_value):
                return min_value
            i += 1
        if(isNegative):
            return -res
        return res
    
	
	
	# 更符合题意的版本，题目要求存储空间是32位的
	def myAtoi(self, str):
		if not str:
			return 0
		i = 0
		size = len(str)
		while i<size and str[i]==' ':
			i += 1
		if i==size:
			return 0
		is_negative = True if str[i]=='-' else False
		res = 0
		if str[i] in ('-','+'):
			i += 1
		while i<size and str[i]>='0' and str[i]<='9':
			n = ord(str[i])-ord('0')
			if is_negative and (-res<-214748364 or (-res==-214748364 and n>=8)):
				return -2147483648
			if not is_negative and (res>214748364 or (res==214748364 and n>=7)):
				return 2147483647
			res = res*10 + n
			i += 1
		return -res if is_negative else res	
	
	